src/test/ui/issue-45157.stderr

```diff
@@ -0,0 +1,20 @@
+error[E0502]: cannot borrow `u.z.c` as immutable because it is also borrowed as mutable
+  --> $DIR/issue-45157.rs:37:20
+   |
+34 |         let mref = &mut u.s.a;
+   |                    ---------- mutable borrow occurs here
+...
+37 |         let nref = &u.z.c;
+   |                    ^^^^^^ immutable borrow occurs here
+
+error[E0502]: cannot borrow `u.s.a` as mutable because it is also borrowed as immutable
+  --> $DIR/issue-45157.rs:39:27
```

Comment:
This second error is a bit surprising. I don't quite understand what it is saying, it looks a bit fishy.

cc @pnkfelix -- agreed?

Reply:
Are you saying that there should be no error here at all? Or just that the error should be focused on prefixes of the field projections it is currently highlighting?

Reply:
I do not think there should be an error at all, and certainly not with these spans. Usually something like this:

```rust
let mut a = 2;
let mref = &mut a;
let nref = &a;
println("{}{}", mref, nref);
```

gives only one error, right?

i.e., one borrow comes first, and it "wins"

Reply:
Example: https://play.rust-lang.org/?gist=695ae34722e06a4060c38df82a45c92d&version=nightly

Reply:
So, looking into this a little bit, here's what I know - this error is reported in the function below:

Reply:
So, looking into this a little bit, here is what I've figured out:

https://github.com/rust-lang/rust/blob/fd0f29237c9562efef25fb30f03df24edb1e2042/src/librustc_mir/util/borrowck_errors.rs#L225-L247

The error is reported in the above function. That is called by the following function:

https://github.com/rust-lang/rust/blob/fdecb0564b86f6876124903a05e590e929532039/src/librustc_mir/borrow_check/error_reporting.rs#L259-L272

This is quite similar to the work done for #47607, in that PR, I added a set that contains the place/span of any errors reported so that they aren't reported again. In this case, the span on line 37 on the below error (that we want) would be in this set:

```
error[E0502]: cannot borrow `u.z.c` as immutable because it is also borrowed as mutable
  --> src/main.rs:37:20
   |
34 |         let mref = &mut u.s.a; //~ ERROR cannot assign twice to immutable variable `mref` [E0384]
   |                    ---------- mutable borrow occurs here
...
37 |         let nref = &u.z.c; //~ ERROR cannot assign twice to immutable variable `err` [E0384]
   |                    ^^^^^^ immutable borrow occurs here
```

I'm not entirely sure what the unintended side effects of the following might be, but we could check if the `issued_span` is in the set (in the above example, that is the error on line 34, but in the second unintended error, that would refer to the same location as above on line 37) and if it is, skip this error. It would essentially silence errors that overlap where the second borrow location of the first error is the first borrow location in subsequent errors. Thoughts?

Reply:
@davidtwco do you think you could gist the output from `-Zdump-mir=nll` for this function?

Reply:
@nikomatsakis is this what you're looking for?

Reply:
yep but I was hoping for a gist :) kind of hard to digest inline...

Reply:
updated your comment =)

Reply:
Here is a [gist](https://gist.github.com/davidtwco/22dafecab494c823e94661743aa966a6) with some logging added in `access_place` and `check_access_for_conflict` that should show the variable values.